modules/common/tls/tls.go

```diff
@@ -0,0 +1,192 @@
+/*
+Copyright 2023 Red Hat
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package tls
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/openstack-k8s-operators/lib-common/modules/common/helper"
+	"github.com/openstack-k8s-operators/lib-common/modules/common/secret"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/util/intstr"
+	"k8s.io/client-go/kubernetes"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/client/config"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+)
+
+// Service contains server-specific TLS secret
+type Service struct {
+	// Server-specific settings
+	SecretName             string `json:"secretName,omitempty"`
+	DisableNonTLSListeners bool   `json:"disableNonTLSListeners,omitempty"`
+}
+
+// Ca contains CA-specific settings, which could be used both by services (to define their own CA certificates)
+// and by clients (to verify the server's certificate)
+type Ca struct {
+	// CA-specific settings
+	CaSecretName string `json:"caSecretName,omitempty"`
+}
+
+// TLS - a generic type, which encapsulates both the service and CA configurations
+type TLS struct {
+	Service Service `json:"service"`
+	Ca      Ca      `json:"ca"`
+}
+
+// NewTLS - initialize and return a TLS struct
+func NewTLS(cr client.Object, clientset *kubernetes.Clientset, scheme *runtime.Scheme, namespace string, service Service, ca Ca) (*TLS, error) {
```

Comment:
should also pass in the context, instead of use `context.TODO()`

Reply:
how should we handle case where it is not a service, and instead only requires the CA volume mount. could use ptr to Service and CA  for and when service is nil we only return the volume mounts for the CA?